randovania/generator/base_patches_factory.py

```diff
@@ -108,7 +108,7 @@ def configurable_node_assignment(self, configuration: BaseConfiguration, game: G
 TrilogyConfiguration = PrimeConfiguration | EchoesConfiguration | CorruptionConfiguration
 
 
-class PrimeTrilogyBasePatchesFactory(BasePatchesFactory):
+class ElevatorRandoBasePatchesFactory(BasePatchesFactory):
```

Comment:
As a "Base for Prime Trilogy" it's fine, but not for a feature. It'd need to be some sort of composition

Reply:
No idea what you mean. It's only renamed to accommodate that is now also used by dread instead of only prime games. The usage is still the same.
If that's an issue, then it shouldn't have been there in the first place as it was only adding the elevator rando feature to prime games with a `if echoes` hack.

Reply:
In practice, this PR is changing code intended only for the three prime games into something usable by any game in Randovania.

I'm okay with weird code and `if game ==` in components that are game specific, but it's very no for anything generic.

Reply:
Maybe I just could not imagine how other games do it or will do it but that does not really hold up for me because game specific and generic is flipped between patcher factory and the elevator tab.

`PrimeTrilogyBasePatchesFactory` is not game specific for me as it is generic for configurations which (at least) implemented a `elevators` field. Which btw isn't the case for `CorruptionConfiguration`, so it shouldn't be there. It's "Prime trilogy without corruption" and prime trilogy is not game specific for me. Besides the name and the type for `TrilogyConfiguration`, the code does not look like it was only intended for prime. The only "not so nice" thing is the `elevator_echoes_shuffled` method mixed into it. Hence, I can not understand the reason why this needs a rewrite

`PresetElevators` is used in a generic way, as it is just throwed into the (prime) games which should use it, but it's not generic as it explicitly needs to check for the game and needs to do all the visibility and text changes. Also the `sound_bug_warning` and `custom_weight` needs to check for the instance / game during runtime. The ui files for prime1, echoes and (deactivated) corruption should be seperated but, once again, that should not have been like that in the first place and did not arose with this pr.

Also what you mean by 
> I'm okay with weird code and `if game ==` in components that are game specific

that sounds like a contradiction as game specific components already know which game "they are" by the pure fact that the line is currently executed. If it needs to check for the game, it's not game specfic.

Reply:
"Prime Trilogy" is game specific. It's for these games and these games alone. The widget and elevator makes assumptions it'll always be about a prime game. Critical example is having the skip final bosses setting be related to elevator rando, because in these games these are entirely related.
Even Corruption, where it doesn't even have elevators in the same way the same assumptions work because the game still has elevators. But everything related to Corruption and Elevator rando is dumb and could/should be removed, it provides no value.

`PrimeTrilogyBasePatchesFactory` code as-is might fit other games, but the purpose of that class was for any shared code for the prime games. I think it had more in the past, but that got refactored away?

> The only "not so nice" thing is the elevator_echoes_shuffled method mixed into it. Hence, I can not understand the reason why this needs a rewrite

It's because it's a bad design overall. Composition would go a long way. It could just be `add_elevator_connections_to_patches` as a standalone function. Even simpler, a thing that takes a `GameDescription`, a `Random` and a `TeleporterConfiguration` and returns a list of `DockConnection` things.


> The ui files for prime1, echoes and (deactivated) corruption should be seperated but, once again, that should not have been like that in the first place and did not arose with this pr.

Don't forget that you're touching code written before 99% of the tooling we have for per-game things didn't exist. The only thing one could use to identify which game it was by checking for the `RandovaniaGame`.
It's not acceptable for Prime 1/2 itself, but since it's just in the corner being used by only these two it's kinda fine. But I'd prefer a proper cleanup if it's going to be used by anything else.

Reply:
> "Prime Trilogy" is game specific. It's for these games and these games alone. The widget and elevator makes assumptions it'll always be about a prime game. Critical example is having the skip final bosses setting be related to elevator rando, because in these games these are entirely related.

In the case of the ui, I disagree with that. skip final boss check is one thing they have common but echoes additionally adds custom weights, sound bug warning and allow unvisted names check. For the tab dread and prime1 are closer together than echoes and prime1.

In my current local WIP, I have a hierarchy of `PresetElevatorsCommon` -> `PresetElevatorsPrimeTrilogy` -> `PresetElevatorsPrime1` or `PresetElevatorsPrime2`
Common is what all shares, Trilogy is what is adding skip final boss check, Prime1 just mixin the UI + removes `ECHOES_SHUFFLED` from `compatible_modes`, Prime2 mixin the ui, custom weights, allow unvisted names check, sound bug warning.
Don't really like that as `PresetElevatorsCommon` needs to make assumptions of what is mixed in via the ui to a child's class. E.g. IDE will complain that there is no such thing as `elevator_layout`, `setupUi` etc. as it's only defined in child classes.

Is there a better way to define the differences + using a different ui for each game while sharing a common logic which is heavily tied to the ui?

> It could just be `add_elevator_connections_to_patches` as a standalone function. Even simpler, a thing that takes a `GameDescription`, a `Random` and a `TeleporterConfiguration` and returns a list of `DockConnection` things.

I have something like `def get_dock_connections_for_elevators(game_description: GameDescription, rng: Random, configuration: GamesWithElevatorConfiguration)` but don't know when to call it.
If I call it within `BasePatchesFactory`'s `add_elevator_connections_to_patches`, I need to check if the game supports it but  there is no flag for that?
If I overwrite `add_elevator_connections_to_patches` in a subclass which is inherited by every game which supports it, it's very much the solution which is already in the current state of this pr.

Reply:
The unvisited names was supposed to be in prime 1 too, but lacks patcher support :(

> Don't really like that as PresetElevatorsCommon needs to make assumptions of what is mixed in via the ui to a child's class.
I'd just have the entire UI be per-game. All that is really being shared is code for creating the checkboxes for each elevator, and even that is going to be different for Dread no?

>  If I call it within BasePatchesFactory's add_elevator_connections_to_patches, I need to check if the game supports it but there is no flag for that?

`BasePatchesFactory` shouldn't have a `add_elevator_connections_to_patches`. It should have a `add_dock_connections_to_patches` that is by default empty. Better yet would be a `dock_connection_assignment`, similar to `configurable_node_assignment`.
Then any game that has elevator rando can just implement that method with `return get_dock_connections_for_elevators(...)`

It's fine to have a bit repeated code in each game, as long as it's not the _exactly_ same code in _every_ game.

Reply:
> I'd just have the entire UI be per-game. All that is really being shared is code for creating the checkboxes for each elevator, and even that is going to be different for Dread no?

That part was well hidden. I pushed the WIP of only the hierarchy. Everything in common is shared among prime1, prime2 and dread. And I have some inner struggle in repeating all this stuff for each Ui file.
BTW: ignore the nice typo `PrimeTriologyTeleporterConfiguration` and that it is used there. it is really "just have a `elevators` field in your configuration"

Reply:
Changed it to use a seperate file + ui for each game even if it hurts me to have that many copy & paste code.